File: Data Structures & Algorithms/validate-parentheses/submission-2.py
class Solution:
  def isValid(self, s: str) -> bool:
    brackets = []
    
    for i in s:
      if i == '(':
        brackets.append('(')
        
      elif i == '[':
        brackets.append('[')
        
      elif i == '{':
        brackets.append('{')
      
      elif i in (')', ']', '}'):
        
        if len(brackets) == 0:
          return False
        
        elif i == ')':
          if brackets.pop() != '(':
            return False
          
        elif i == ']':
          if brackets.pop() != '[':
            return False
          
        elif i == '}':
          if brackets.pop() != '{':
            return False
        
    return len(brackets) == 0

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
      Stack<TreeNode> stack = new Stack<>();
      Map<TreeNode, int[]> map = new HashMap<>();
      map.put(null, new int[] {0, 0});
      stack.push(root);

      while (!stack.empty()) {
        var current = stack.peek();

        if (!map.containsKey(current.left) && current.left != null) {
          stack.push(current.left);
        } else if (!map.containsKey(current.right) && current.right != null) {
          stack.push(current.right);
        } else {
          var leftPair = map.get(current.left);
          var rightPair = map.get(current.right);

          int height = Math.max(leftPair[0], rightPair[0]) + 1;
          int width = Math.max(
            Math.max(leftPair[1], rightPair[1]),
            leftPair[0] + rightPair[0]
          );

          map.put(current, new int[] {height, width});
          stack.pop();
        }
      }



      return map.get(root)[1];
    }
}
